File: Sources/Buckett/Services/SigV4.swift
import Foundation
import CryptoKit

/// Minimal AWS Signature Version 4 signer — enough to talk to any S3-compatible API
/// (Cloudflare R2, Backblaze B2, AWS S3, MinIO, …) without pulling in an SDK.
enum SigV4 {

    struct Credentials {
        var accessKeyID: String
        var secretAccessKey: String
    }

    static let unreservedCharacters = CharacterSet(
        charactersIn: "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-._~"
    )
    static let unreservedWithSlash = unreservedCharacters.union(CharacterSet(charactersIn: "/"))

    /// RFC 3986 percent-encoding as required by SigV4.
    static func encode(_ string: String, keepSlash: Bool = false) -> String {
        string.addingPercentEncoding(
            withAllowedCharacters: keepSlash ? unreservedWithSlash : unreservedCharacters
        ) ?? string
    }

    static func sha256Hex(_ data: Data) -> String {
        SHA256.hash(data: data).map { String(format: "%02x", $0) }.joined()
    }

    static let emptyPayloadHash = sha256Hex(Data())

    private static func hmac(key: Data, data: Data) -> Data {
        Data(HMAC<SHA256>.authenticationCode(for: data, using: SymmetricKey(data: key)))
    }

    private static func hmac(key: Data, string: String) -> Data {
        hmac(key: key, data: Data(string.utf8))
    }

    private static let amzDateFormatter: DateFormatter = {
        let f = DateFormatter()
        f.dateFormat = "yyyyMMdd'T'HHmmss'Z'"
        f.timeZone = TimeZone(identifier: "UTC")
        f.locale = Locale(identifier: "en_US_POSIX")
        return f
    }()

    /// Signs `request` in place. The request URL's percent-encoded path and query must
    /// already be in canonical (SigV4) form — `S3Client` builds them that way.
    static func sign(
        request: inout URLRequest,
        payloadHash: String,
        credentials: Credentials,
        region: String,
        service: String = "s3",
        date: Date = Date()
    ) {
        guard let url = request.url,
              let components = URLComponents(url: url, resolvingAgainstBaseURL: false),
              let host = components.host
        else { return }

        let amzDate = amzDateFormatter.string(from: date)
        let dateStamp = String(amzDate.prefix(8))

        var hostHeader = host
        if let port = components.port, port != 443, port != 80 {
            hostHeader += ":\(port)"
        }

        request.setValue(amzDate, forHTTPHeaderField: "x-amz-date")
        request.setValue(payloadHash, forHTTPHeaderField: "x-amz-content-sha256")

        // Headers included in the signature: host, content-md5/content-type when present,
        // and every x-amz-* header on the request.
        var headersToSign: [String: String] = ["host": hostHeader]
        for (name, value) in request.allHTTPHeaderFields ?? [:] {
            let lower = name.lowercased()
            if lower.hasPrefix("x-amz-") || lower == "content-md5" || lower == "content-type" {
                headersToSign[lower] = value.trimmingCharacters(in: .whitespaces)
            }
        }

        let sortedNames = headersToSign.keys.sorted()
        let canonicalHeaders = sortedNames.map { "\($0):\(headersToSign[$0]!)\n" }.joined()
        let signedHeaders = sortedNames.joined(separator: ";")

        let canonicalURI = components.percentEncodedPath.isEmpty ? "/" : components.percentEncodedPath
        let canonicalQuery = components.percentEncodedQuery ?? ""

        let canonicalRequest = [
            request.httpMethod ?? "GET",
            canonicalURI,
            canonicalQuery,
            canonicalHeaders,
            signedHeaders,
            payloadHash
        ].joined(separator: "\n")

        let scope = "\(dateStamp)/\(region)/\(service)/aws4_request"
        let stringToSign = [
            "AWS4-HMAC-SHA256",
            amzDate,
            scope,
            sha256Hex(Data(canonicalRequest.utf8))
        ].joined(separator: "\n")

        let kDate = hmac(key: Data(("AWS4" + credentials.secretAccessKey).utf8), string: dateStamp)
        let kRegion = hmac(key: kDate, string: region)
        let kService = hmac(key: kRegion, string: service)
        let kSigning = hmac(key: kService, string: "aws4_request")
        let signature = hmac(key: kSigning, string: stringToSign)
            .map { String(format: "%02x", $0) }.joined()

        let authorization = "AWS4-HMAC-SHA256 " +
            "Credential=\(credentials.accessKeyID)/\(scope), " +
            "SignedHeaders=\(signedHeaders), " +
            "Signature=\(signature)"
        request.setValue(authorization, forHTTPHeaderField: "Authorization")
    }
}

File: windows/src/Buckett/Services/SigV4.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Linq;
using System.Net.Http;
using System.Security.Cryptography;
using System.Text;

namespace Buckett.Services;

/// Minimal AWS Signature Version 4 signer — enough to talk to any S3-compatible API
/// (Cloudflare R2, Backblaze B2, AWS S3, MinIO, …) without pulling in an SDK.
public static class SigV4
{
    public readonly record struct Credentials(string AccessKeyID, string SecretAccessKey);

    private const string Unreserved =
        "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-._~";

    /// RFC 3986 percent-encoding as required by SigV4.
    public static string Encode(string value, bool keepSlash = false)
    {
        var builder = new StringBuilder(value.Length + 16);
        foreach (var b in Encoding.UTF8.GetBytes(value))
        {
            var c = (char)b;
            if (Unreserved.IndexOf(c) >= 0 || (keepSlash && c == '/'))
            {
                builder.Append(c);
            }
            else
            {
                builder.Append('%').Append(b.ToString("X2", CultureInfo.InvariantCulture));
            }
        }
        return builder.ToString();
    }

    public static string Sha256Hex(ReadOnlySpan<byte> data) =>
        Convert.ToHexString(SHA256.HashData(data)).ToLowerInvariant();

    public static string Sha256Hex(string value) => Sha256Hex(Encoding.UTF8.GetBytes(value));

    public static readonly string EmptyPayloadHash = Sha256Hex(ReadOnlySpan<byte>.Empty);

    private static byte[] Hmac(byte[] key, string data) =>
        HMACSHA256.HashData(key, Encoding.UTF8.GetBytes(data));

    public static string AmzDate(DateTime utc) =>
        utc.ToString("yyyyMMdd'T'HHmmss'Z'", CultureInfo.InvariantCulture);

    private static byte[] SigningKey(string secret, string dateStamp, string region, string service)
    {
        var kDate = Hmac(Encoding.UTF8.GetBytes("AWS4" + secret), dateStamp);
        var kRegion = Hmac(kDate, region);
        var kService = Hmac(kRegion, service);
        return Hmac(kService, "aws4_request");
    }

    /// Builds a presigned GET/PUT URL (SigV4 query-string auth) valid for `expires` seconds.
    /// `canonicalPath` must already be in canonical percent-encoded form.
    public static string? Presign(
        string scheme,
        string host,
        int port,
        string canonicalPath,
        Credentials credentials,
        string region,
        int expires,
        string method = "GET",
        string service = "s3",
        DateTime? date = null)
    {
        var utc = (date ?? DateTime.UtcNow).ToUniversalTime();
        var amzDate = AmzDate(utc);
        var dateStamp = amzDate[..8];
        var scope = $"{dateStamp}/{region}/{service}/aws4_request";

        var hostHeader = host;
        if (port != 443 && port != 80 && port > 0) hostHeader += ":" + port;

        var query = new List<(string Name, string Value)>
        {
            ("X-Amz-Algorithm", "AWS4-HMAC-SHA256"),
            ("X-Amz-Credential", $"{credentials.AccessKeyID}/{scope}"),
            ("X-Amz-Date", amzDate),
            ("X-Amz-Expires", expires.ToString(CultureInfo.InvariantCulture)),
            ("X-Amz-SignedHeaders", "host")
        };

        var canonicalQuery = string.Join("&", query
            .Select(pair => (Name: Encode(pair.Name), Value: Encode(pair.Value)))
            .OrderBy(pair => pair.Name, StringComparer.Ordinal)
            .Select(pair => $"{pair.Name}={pair.Value}"));

        var canonicalUri = canonicalPath.Length == 0 ? "/" : canonicalPath;
        var canonicalRequest = string.Join("\n",
            method,
            canonicalUri,
            canonicalQuery,
            $"host:{hostHeader}\n",
            "host",
            "UNSIGNED-PAYLOAD");

        var stringToSign = string.Join("\n",
            "AWS4-HMAC-SHA256",
            amzDate,
            scope,
            Sha256Hex(canonicalRequest));

        var signature = Convert
            .ToHexString(Hmac(SigningKey(credentials.SecretAccessKey, dateStamp, region, service), stringToSign))
            .ToLowerInvariant();

        query.Add(("X-Amz-Signature", signature));
        var finalQuery = string.Join("&", query.Select(pair => $"{Encode(pair.Name)}={Encode(pair.Value)}"));

        var portPart = port is 443 or 80 or <= 0 ? "" : ":" + port;
        return $"{scheme}://{host}{portPart}{canonicalUri}?{finalQuery}";
    }

    /// Signs `request` in place. The request URI's percent-encoded path and query must
    /// already be in canonical (SigV4) form — `S3Client` builds them that way.
    public static void Sign(
        HttpRequestMessage request,
        string canonicalPath,
        string canonicalQuery,
        string payloadHash,
        Credentials credentials,
        string region,
        string service = "s3",
        DateTime? date = null)
    {
        var uri = request.RequestUri;
        if (uri == null) return;

        var utc = (date ?? DateTime.UtcNow).ToUniversalTime();
        var amzDate = AmzDate(utc);
        var dateStamp = amzDate[..8];

        var hostHeader = uri.Host;
        if (!uri.IsDefaultPort) hostHeader += ":" + uri.Port;

        request.Headers.TryAddWithoutValidation("x-amz-date", amzDate);
        request.Headers.TryAddWithoutValidation("x-amz-content-sha256", payloadHash);

        // Headers included in the signature: host, content-md5/content-type when present,
        // and every x-amz-* header on the request.
        var headersToSign = new SortedDictionary<string, string>(StringComparer.Ordinal)
        {
            ["host"] = hostHeader
        };

        void Collect(IEnumerable<KeyValuePair<string, IEnumerable<string>>> headers)
        {
            foreach (var header in headers)
            {
                var lower = header.Key.ToLowerInvariant();
                if (lower.StartsWith("x-amz-", StringComparison.Ordinal) ||
                    lower is "content-md5" or "content-type")
                {
                    headersToSign[lower] = string.Join(",", header.Value).Trim();
                }
            }
        }

        Collect(request.Headers);
        if (request.Content != null) Collect(request.Content.Headers);

        var canonicalHeaders = string.Concat(headersToSign.Select(kv => $"{kv.Key}:{kv.Value}\n"));
        var signedHeaders = string.Join(";", headersToSign.Keys);

        var canonicalUri = canonicalPath.Length == 0 ? "/" : canonicalPath;
        var canonicalRequest = string.Join("\n",
            request.Method.Method,
            canonicalUri,
            canonicalQuery,
            canonicalHeaders,
            signedHeaders,
            payloadHash);

        var scope = $"{dateStamp}/{region}/{service}/aws4_request";
        var stringToSign = string.Join("\n",
            "AWS4-HMAC-SHA256",
            amzDate,
            scope,
            Sha256Hex(canonicalRequest));

        var signature = Convert
            .ToHexString(Hmac(SigningKey(credentials.SecretAccessKey, dateStamp, region, service), stringToSign))
            .ToLowerInvariant();

        var authorization =
            "AWS4-HMAC-SHA256 " +
            $"Credential={credentials.AccessKeyID}/{scope}, " +
            $"SignedHeaders={signedHeaders}, " +
            $"Signature={signature}";
        request.Headers.TryAddWithoutValidation("Authorization", authorization);
    }
}
